hydra-main/src/main/java/com/addthis/hydra/job/web/JobRequestHandlerImpl.java

```diff
@@ -116,6 +115,70 @@ public Job createOrUpdateJob(KVPairs kv, String username, String token, String s
         return job;
     }
 
+    private User checkUserCredentials(String username, String token) {
+        User user = spawn.getPermissionsManager().authenticate(username, token);
+        if (user == null) {
+            throw new InsufficientPrivilegesException(username, "invalid credentials provided");
+        }
+        return user;
+    }
+
+    private void checkJobWritePrivileges(Job job, String username, String token, String sudo) {
+        if (!spawn.getPermissionsManager().isWritable(username, token, sudo, job)) {
+            log.warn("User {} (sudo = {}) had insufficient privileges to modify job {}", username, sudo != null, job.getId());
+            throw new InsufficientPrivilegesException(username, "insufficient privileges to modify job " + job.getId());
+        }
+    }
+
+    @Override
+    public Job updateMinionType(String jobId, String minionType, String username, String token, String sudo)
+            throws Exception {
+        checkUserCredentials(username, token);
+        spawn.acquireJobLock();
+        try {
+            Job job = spawn.getJob(jobId);
+            if (job != null) {
+                checkJobWritePrivileges(job, username, token, sudo);
+                checkMinionTypeUpdatePrerequisite(job, job.getState() == JobState.IDLE, "Job is not idle");
+                checkMinionTypeUpdatePrerequisite(job, isJobOnMinionType(job, minionType),
+                                                  "Not all tasks and replicas are on " + minionType);
+                job.setMinionType(minionType);
+                spawn.updateJob(job);
+            }
+            return job;
+        } finally {
+            spawn.releaseJobLock();
+        }
+    }
+
+    private void checkMinionTypeUpdatePrerequisite(Job job, boolean condition, String errorMessage) {
+        checkArgument(condition, "Cannot update minion type of job " + job.getId() + ": " + errorMessage);
+    }
+
+    private boolean isJobOnMinionType(Job job, String minionType) {
+        // make a valid host set
+        Set<String> hostsWithTargetMinionType = new HashSet<String>();
+        List<HostState> hostStates = spawn.hostManager.listHostStatus(minionType);
+        for (HostState hostState : hostStates) {
+            hostsWithTargetMinionType.add(hostState.getHostUuid());
+        }
+
+        // check tasks and their replicas
+        for (JobTask jobTask : job.getCopyOfTasks()) {
+            if (!hostsWithTargetMinionType.contains(jobTask.getHostUUID())) {
```

Comment:
I read the code, but did not fully understand weather it is the case. So I ran a remote debugger for a job with `number of replicas` set to 2, in which case there are 1+2 copies in total. The debugger shows there are two replicas. I think this means `jobTask.getHostUUID` is not in `jobTask.getReplicas`. I will further read the code to confirm.